### ISSUE TITLE
fsdp grad norm need to be aggregated manually

### ISSUE THREAD

**Issue by gameofdimension** [pytorch/torchtitan#766]
![image](https://github.com/user-attachments/assets/3263a50f-1e30-439b-a1ca-98a507d51671)

I noticed that the `grad_norm` varies across different ranks when applying fsdp, suggesting that gradients are not being aggregated at a global level.

Manual reduction of gradients across all ranks is necessary, similar to the approach taken for PP.

https://github.com/pytorch/torchtitan/blob/3f20451bc3b0de6d292946b1774581856014a0a9/torchtitan/utils.py#L387

**Comment by gameofdimension** [comment]
This is my code to produce the above log

```diff
diff --git a/train.py b/train.py
index 2874d0d..02943bb 100644
--- a/train.py
+++ b/train.py
@@ -308,7 +308,7 @@ def main(job_config: JobConfig):
                     loss.backward()
 
             # clip gradients
-            utils.clip_grad_norm_(
+            grad_norm = utils.clip_grad_norm_(
                 [p for m in model_parts for p in m.parameters()],
                 job_config.training.max_norm,
                 foreach=True,
@@ -390,6 +390,7 @@ def main(job_config: JobConfig):
                 logger.info(
                     f"{color.cyan}step: {train_state.step:2}  "
                     f"{color.green}loss: {global_avg_loss:7.4f}  "
+                    f"{color.green}grad_norm: {grad_norm.item():7.4f}  "
                     f"{color.yellow}memory: {device_mem_stats.max_reserved_gib:5.2f}GiB"
                     f"({device_mem_stats.max_reserved_pct:.2f}%)  "
                     f"{color.blue}tps: {round(tps):,}  "
```

**Comment by awgu** [comment]
Can you try `grad_norm.full_tensor().item()`?

**Comment by gameofdimension** [comment]
It is identical across ranks now. So it suggests that we should also call `total_norm = total_norm.full_tensor()` for fsdp now? 

https://github.com/pytorch/torchtitan/blob/3f20451bc3b0de6d292946b1774581856014a0a9/torchtitan/utils.py#L387-L394

**Comment by awgu** [comment]
cc: @tianyu-l @wanchaol The issue here is that inside `nn.utils.clip_grad_norm_`, the `total_norm` will be `_NormPartial` until it gets used in `clip_coef = max_norm / (total_norm + 1e-6)`([code](https://github.com/pytorch/pytorch/blob/dec1a6d0f05f838dcec10492ef6091501258f816/torch/nn/utils/clip_grad.py#L155)), where it gets converted to `Replicate`.

This is an issue because then if the user wants to log the `total_norm` as `Replicate` in the training loop like in this case, then we will incur the `_NormPartial` -> `Replicate` redistribute twice, which incurs redundant all-reduce.

---

@gameofdimension For torchtitan specifically, since we have forked the `clip_grad_norm_` implementation, I think we could consider some modification, but if we want to avoid the extra all-reduce from redundantly redistributing, we cannot simply call `.full_tensor()` -- we need to pass the replicated `total_norm` into the clipping part too.

**Comment by gameofdimension** [comment]
So it is only a redundant all-reduce, will not cause any error. `clip_grad_norm_` here works as expected?

**Comment by awgu** [comment]
yes

**Comment by gameofdimension** [comment]
> yes

Thank you for your clarification and patience.

**Comment by gameofdimension** [comment]
what if we have both sharded and unsharded parameters? the `total_norm` will be still computed locally and globally?

**Comment by awgu** [comment]
> what if we have both sharded and unsharded parameters? the `total_norm` will be still computed locally and globally? will the unsharded part be accumulated for more than once?

How are your unsharded parameters managed? If they have `Replicate()` placement with `DTensor`, then they will be handled correctly. There should not be any overcounting.

**Comment by tianyu-l** [comment]
@gameofdimension 
Does this address the issue?
https://github.com/pytorch/torchtitan/pull/769

**Comment by gameofdimension** [comment]
> > what if we have both sharded and unsharded parameters? the `total_norm` will be still computed locally and globally? will the unsharded part be accumulated for more than once?
> 
> How are your unsharded parameters managed? If they have `Replicate()` placement with `DTensor`, then they will be handled correctly. There should not be any overcounting.

thanks

**Comment by gameofdimension** [comment]
> @gameofdimension Does this address the issue? #769

👌
